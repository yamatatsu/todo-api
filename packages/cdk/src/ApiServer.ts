import {
  App,
  Stack,
  StackProps,
  aws_ec2,
  aws_lambda,
  aws_lambda_nodejs,
  aws_apigateway,
  aws_secretsmanager,
  Duration,
  aws_cognito,
} from "aws-cdk-lib";

type Props = StackProps & {
  codeEntry: string;
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  dbCredentialSecret: aws_secretsmanager.ISecret;
  proxyEndpoint: string;
  userPool: aws_cognito.IUserPool;
};

export class ApiServerStack extends Stack {
  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    // TODO: lambda insight入れる
    const handler = new aws_lambda_nodejs.NodejsFunction(this, "Lambda", {
      entry: props.codeEntry,
      runtime: aws_lambda.Runtime.NODEJS_14_X,
      // TODO: aurora serverless やめて rds proxyつけたら10秒に戻す
      timeout: Duration.seconds(20),
      vpc: props.vpc,
      securityGroups: [props.securityGroup],
      bundling: {
        /**
         * esbuildの成果物にprismaのengineが含める必要がある
         * Respect for https://dev.to/prisma/bundling-prisma-with-the-cdk-using-aws-lambda-nodejs-2lkd
         */
        nodeModules: ["@prisma/client", "prisma"],
        commandHooks: {
          beforeBundling(inputDir: string, outputDir: string): string[] {
            return [];
          },
          beforeInstall(inputDir: string, outputDir: string) {
            return [`cp -R ../prisma ${outputDir}/`];
          },
          afterBundling(inputDir: string, outputDir: string): string[] {
            return [
              `cd ${outputDir}`,
              `yarn prisma generate`,
              `rm -rf node_modules/@prisma/engines`,
            ];
          },
        },
        /**
         * forceDockerBundling: これは default false であるが、下記理由により明示的に示す。
         * これが true のとき、もしくは esbuild が見つからないとき(下記URL参照)に、
         * NodejsFunction では docker を用いてビルドが行われる。
         * https://github.com/aws/aws-cdk/blob/4c8e938e01b87636390a4f04de63bcd4dfe44cf8/packages/@aws-cdk/aws-lambda-nodejs/lib/esbuild-installation.ts#L8-L31
         * dockerでビルドする場合、特に Mac M1 での実行結果と GitHub actions 上での実行結果とで、
         * 成果物のfingerprintがズレて snapshot test が落ちてしまう。
         * そのため、 forceDockerBundling を明示し、 package.json でも esbuild を明示的にインストールする。
         */
        forceDockerBundling: false,
      },
      environment: {
        NODE_ENV: "production",
        SECRET_NAME: props.dbCredentialSecret.secretName,
        DATABASE_ENDPOINT: props.proxyEndpoint,
      },
    });

    props.dbCredentialSecret.grantRead(handler);

    // const authorizer = new aws_apigateway.CognitoUserPoolsAuthorizer(
    //   this,
    //   "Authorizer",
    //   {
    //     cognitoUserPools: [],
    //     identitySource: "method.request.header.x-ta-token",
    //   }
    // );

    const restApi = new aws_apigateway.LambdaRestApi(this, "LambdaRestApi", {
      handler,
      deployOptions: {
        metricsEnabled: true,
        loggingLevel: aws_apigateway.MethodLoggingLevel.INFO,
        dataTraceEnabled: true,
        tracingEnabled: true,
      },
      // defaultMethodOptions: {
      //   authorizer,
      // },
    });
  }
}
