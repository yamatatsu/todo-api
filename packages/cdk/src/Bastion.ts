import { App, Stack, StackProps, aws_ec2, aws_iam } from "aws-cdk-lib";

type BastionProps = {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
} & StackProps;

export class BastionStack extends Stack {
  readonly bastion: aws_ec2.Instance;

  constructor(scope: App, id: string, props: BastionProps) {
    super(scope, id, props);

    const instanceType = new aws_ec2.InstanceType("t2.micro");
    const machineImage = new aws_ec2.AmazonLinuxImage({
      generation: aws_ec2.AmazonLinuxGeneration.AMAZON_LINUX_2,
    });

    // ec2 instance for bastion server.
    this.bastion = new aws_ec2.Instance(this, "Bastion", {
      instanceType: instanceType,
      machineImage: machineImage,
      vpc: props.vpc,
      vpcSubnets: { subnetGroupName: "application" },
    });

    this.bastion.addSecurityGroup(props.securityGroup);

    // for ssm session manager
    this.bastion.addToRolePolicy(
      new aws_iam.PolicyStatement({
        effect: aws_iam.Effect.ALLOW,
        actions: [
          "ssm:UpdateInstanceInformation",
          "ssmmessages:CreateControlChannel",
          "ssmmessages:CreateDataChannel",
          "ssmmessages:OpenControlChannel",
          "ssmmessages:OpenDataChannel",
        ],
        resources: ["*"],
      })
    );

    this.bastion.addToRolePolicy(
      new aws_iam.PolicyStatement({
        effect: aws_iam.Effect.ALLOW,
        actions: ["s3:GetEncryptionConfiguration"],
        resources: ["*"],
      })
    );
  }
}
