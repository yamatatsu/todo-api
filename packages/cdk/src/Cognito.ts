import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_cognito,
} from "aws-cdk-lib";

export class CognitoStack extends Stack {
  public readonly userPool: aws_cognito.IUserPool;

  constructor(parent: App, id: string, props?: StackProps) {
    super(parent, id, props);

    const userPool = new aws_cognito.UserPool(this, "UserPool", {
      selfSignUpEnabled: true,
      signInAliases: { username: true, email: true },
      customAttributes: {
        validated_email: new aws_cognito.StringAttribute({
          mutable: true,
        }),
      },
      passwordPolicy: {
        // Because of https://pages.nist.gov/800-63-3/sp800-63b.html#-5112-memorized-secret-verifiers
        minLength: 8,
        // Because of https://pages.nist.gov/800-63-3/sp800-63b.html#a3-complexity
        requireLowercase: false,
        requireUppercase: false,
        requireDigits: false,
        requireSymbols: false,
      },
      removalPolicy: RemovalPolicy.DESTROY,
    });

    new aws_cognito.UserPoolClient(this, "UserPoolClient", {
      userPool,
      authFlows: { userSrp: true },
    });

    this.userPool = userPool;
  }
}
