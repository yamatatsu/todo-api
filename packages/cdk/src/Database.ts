import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_ec2,
  aws_rds,
  aws_secretsmanager,
} from "aws-cdk-lib";

type Props = StackProps & {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
};

export class DatabaseStack extends Stack {
  public readonly dbCredentialSecret: aws_secretsmanager.ISecret;

  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const dbCluster = new aws_rds.ServerlessCluster(this, "Aurora", {
      engine: aws_rds.DatabaseClusterEngine.AURORA,
      credentials: aws_rds.Credentials.fromGeneratedSecret("LambdaRdsProxy"),
      defaultDatabaseName: "tadb",
      vpc: props.vpc,
      vpcSubnets: { subnetType: aws_ec2.SubnetType.ISOLATED },
      securityGroups: [props.securityGroup],
      removalPolicy: RemovalPolicy.DESTROY,
    });

    // production codeではないので、カジュアルに`!`使う
    this.dbCredentialSecret = dbCluster.secret!;
  }
}
