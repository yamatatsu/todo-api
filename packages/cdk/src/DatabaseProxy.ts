import {
  App,
  Stack,
  StackProps,
  RemovalPolicy,
  aws_ec2,
  aws_rds,
  aws_secretsmanager,
} from "aws-cdk-lib";

type Props = StackProps & {
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  cluster: aws_rds.IDatabaseCluster;
  secret: aws_secretsmanager.ISecret;
};

export class DatabaseProxyStack extends Stack {
  public readonly proxy: aws_rds.IDatabaseProxy;

  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const proxy = props.cluster.addProxy("RdsProxy", {
      vpc: props.vpc,
      secrets: [props.secret],
      securityGroups: [props.securityGroup],
    });
    // proxy.connections.allowFromAnyIpv4(aws_ec2.Port.tcp(3306));

    this.proxy = proxy;
  }
}
