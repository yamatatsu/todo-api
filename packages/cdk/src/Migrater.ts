import {
  App,
  Stack,
  StackProps,
  aws_ec2,
  aws_lambda,
  aws_secretsmanager,
  Duration,
} from "aws-cdk-lib";

type Props = StackProps & {
  dockerfilePath: string;
  vpc: aws_ec2.IVpc;
  securityGroup: aws_ec2.ISecurityGroup;
  dbCredentialSecret: aws_secretsmanager.ISecret;
  proxyEndpoint: string;
};

/**
 * Database に対して prisma migrate deploy をかけるための lambda 関数を定義する。
 * application subnet 内で実行されることで、databaseへアクセス可能となる。
 */
export class MigraterStack extends Stack {
  constructor(parent: App, id: string, props: Props) {
    super(parent, id, props);

    const handler = new aws_lambda.DockerImageFunction(this, "Lambda", {
      functionName: `${id}-Fn`,
      code: aws_lambda.DockerImageCode.fromImageAsset(props.dockerfilePath),
      // 途中で実行止まると怖いので5分くらいを設定する
      timeout: Duration.minutes(5),
      vpc: props.vpc,
      vpcSubnets: { subnetGroupName: "application" },
      securityGroups: [props.securityGroup],
      environment: {
        SECRET_NAME: props.dbCredentialSecret.secretName,
        DATABASE_ENDPOINT: props.proxyEndpoint,
      },
    });

    props.dbCredentialSecret.grantRead(handler);
  }
}
