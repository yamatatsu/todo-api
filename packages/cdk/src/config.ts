const ENV_NAME = ["development", "production"] as const;
type EnvName = typeof ENV_NAME[number];

export function assertEnvName(str: string): asserts str is EnvName {
  if (!ENV_NAME.includes(str as EnvName)) {
    throw new Error(`Invalid EnvName has set: ${str}`);
  }
}

export const AccountId: Record<EnvName, string> = {
  development: "407172421073",
  production: "",
};
