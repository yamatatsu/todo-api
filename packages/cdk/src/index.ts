import path from "path";
import { App, Environment } from "aws-cdk-lib";
import { CognitoStack } from "./Cognito";
import { VpcStack } from "./Vpc";
import { DatabaseStack } from "./Database";
import { ApiServerStack } from "./ApiServer";
import { BastionStack } from "./Bastion";
import { AccountId, assertEnvName } from "./config";

const envName = process.env.ENV_NAME || "development";
assertEnvName(envName);

const app = new App();
const env: Environment = {
  region: "ap-northeast-1",
  account: AccountId[envName],
};

const cognito = new CognitoStack(app, `${envName}-TodoApi-Cognito`, { env });
const { vpc, dbSG, dbAccessSG } = new VpcStack(app, `${envName}-TodoApi-Vpc`, {
  env,
});
const database = new DatabaseStack(app, `${envName}-TodoApi-Database`, {
  vpc,
  securityGroup: dbSG,
  env,
});
new ApiServerStack(app, `${envName}-TodoApi-ApiServer`, {
  codeEntry: path.resolve(__dirname, "../../server/src/index.ts"),
  vpc,
  securityGroup: dbAccessSG,
  dbCredentialSecret: database.dbCredentialSecret,
  userPoolArn: cognito.userPoolArn,
  env,
});
new BastionStack(app, `${envName}-TodoApi-Bastion`, {
  vpc,
  securityGroup: dbAccessSG,
  env,
});
