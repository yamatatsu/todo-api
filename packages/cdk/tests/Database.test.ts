import { SynthUtils } from "@aws-cdk/assert";
import { App, Stack, Environment, aws_ec2 } from "aws-cdk-lib";
import { DatabaseStack } from "../src/Database";
import { AccountId } from "../src/config";

test("snapshot test", () => {
  const env: Environment = {
    region: "ap-northeast-1",
    account: AccountId["development"],
  };

  const app = new App();
  const stack = new Stack(app, "dummyStack", { env });

  const vpc = new aws_ec2.Vpc(stack, "Vpc", {
    subnetConfiguration: [
      { name: "ingress", subnetType: aws_ec2.SubnetType.PUBLIC },
      { name: "application", subnetType: aws_ec2.SubnetType.PRIVATE },
      { name: "rds", subnetType: aws_ec2.SubnetType.ISOLATED },
    ],
  });

  const target = new DatabaseStack(app, "Target", { vpc, env });

  expect(SynthUtils.toCloudFormation(target)).toMatchSnapshot();
});
