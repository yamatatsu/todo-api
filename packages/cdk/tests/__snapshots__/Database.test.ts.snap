// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "Resources": Object {
    "Aurora2CBAB212": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBSubnetGroupName": Object {
          "Ref": "AuroraSubnetsC4DF45C9",
        },
        "DatabaseName": "tadb",
        "Engine": "aurora",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "TargetAuroraSecret852055C73fdaad7efa858a3daf9490cf0a702aeb",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": "dbAdmin",
        "VpcSecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "DBSG9487F3EF",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBCluster",
      "UpdateReplacePolicy": "Delete",
    },
    "AuroraInstance10510C5D1": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBClusterIdentifier": Object {
          "Ref": "Aurora2CBAB212",
        },
        "DBInstanceClass": "db.t3.medium",
        "DBSubnetGroupName": Object {
          "Ref": "AuroraSubnetsC4DF45C9",
        },
        "Engine": "aurora",
        "PubliclyAccessible": false,
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Delete",
    },
    "AuroraInstance2BC8C1D03": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBClusterIdentifier": Object {
          "Ref": "Aurora2CBAB212",
        },
        "DBInstanceClass": "db.t3.medium",
        "DBSubnetGroupName": Object {
          "Ref": "AuroraSubnetsC4DF45C9",
        },
        "Engine": "aurora",
        "PubliclyAccessible": false,
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Delete",
    },
    "AuroraRdsProxyB8FEFA0E": Object {
      "Properties": Object {
        "Auth": Array [
          Object {
            "AuthScheme": "SECRETS",
            "IAMAuth": "DISABLED",
            "SecretArn": Object {
              "Ref": "AuroraSecretAttachmentEAAB0558",
            },
          },
        ],
        "DBProxyName": "RdsProxy",
        "EngineFamily": "MYSQL",
        "RequireTLS": false,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "AuroraRdsProxyIAMRole6EFC0823",
            "Arn",
          ],
        },
        "VpcSecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "DBSG9487F3EF",
              "GroupId",
            ],
          },
        ],
        "VpcSubnetIds": Array [
          Object {
            "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcapplicationSubnet1Subnet8B68205B1F66BB2C",
          },
          Object {
            "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcapplicationSubnet2Subnet296709622ABA6FC7",
          },
          Object {
            "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcapplicationSubnet3SubnetB937EB65D0FA57B0",
          },
        ],
      },
      "Type": "AWS::RDS::DBProxy",
    },
    "AuroraRdsProxyIAMRole6EFC0823": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "rds.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "AuroraRdsProxyIAMRoleDefaultPolicy221FE317": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "AuroraSecretAttachmentEAAB0558",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AuroraRdsProxyIAMRoleDefaultPolicy221FE317",
        "Roles": Array [
          Object {
            "Ref": "AuroraRdsProxyIAMRole6EFC0823",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AuroraRdsProxyProxyTargetGroup475E511C": Object {
      "DependsOn": Array [
        "AuroraInstance10510C5D1",
        "AuroraInstance2BC8C1D03",
        "AuroraRdsProxyIAMRoleDefaultPolicy221FE317",
        "AuroraRdsProxyIAMRole6EFC0823",
        "AuroraRdsProxyB8FEFA0E",
        "Aurora2CBAB212",
        "AuroraSecretAttachmentEAAB0558",
        "TargetAuroraSecret852055C73fdaad7efa858a3daf9490cf0a702aeb",
        "AuroraSubnetsC4DF45C9",
      ],
      "Properties": Object {
        "ConnectionPoolConfigurationInfo": Object {},
        "DBClusterIdentifiers": Array [
          Object {
            "Ref": "Aurora2CBAB212",
          },
        ],
        "DBProxyName": Object {
          "Ref": "AuroraRdsProxyB8FEFA0E",
        },
        "TargetGroupName": "default",
      },
      "Type": "AWS::RDS::DBProxyTargetGroup",
    },
    "AuroraSecretAttachmentEAAB0558": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "TargetAuroraSecret852055C73fdaad7efa858a3daf9490cf0a702aeb",
        },
        "TargetId": Object {
          "Ref": "Aurora2CBAB212",
        },
        "TargetType": "AWS::RDS::DBCluster",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "AuroraSubnetsC4DF45C9": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnets for Aurora database",
        "SubnetIds": Array [
          Object {
            "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcrdsSubnet1SubnetC2926CEA415B78A7",
          },
          Object {
            "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcrdsSubnet2Subnet70A835C80F5EB08B",
          },
          Object {
            "Fn::ImportValue": "dummyStack:ExportsOutputRefVpcrdsSubnet3SubnetAA566B5A902489E7",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "DBAccessSGBB77343E": Object {
      "Properties": Object {
        "GroupDescription": "for accessing database",
        "GroupName": "Database Access",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "dummyStack:ExportsOutputRefVpc8378EB38272D6E3A",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DBSG9487F3EF": Object {
      "Properties": Object {
        "GroupDescription": "for database",
        "GroupName": "Database",
        "VpcId": Object {
          "Fn::ImportValue": "dummyStack:ExportsOutputRefVpc8378EB38272D6E3A",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DBSGfromTargetDBAccessSGA573EF9C33062F29BB9A": Object {
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "from application with sg named ",
              Object {
                "Ref": "DBAccessSGBB77343E",
              },
            ],
          ],
        },
        "FromPort": 3306,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "DBSG9487F3EF",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "DBAccessSGBB77343E",
            "GroupId",
          ],
        },
        "ToPort": 3306,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "DBSGfromTargetDBSG2D7B6556IndirectPort60B97122": Object {
      "Properties": Object {
        "Description": "Allow connections to the database Cluster from the Proxy",
        "FromPort": Object {
          "Fn::GetAtt": Array [
            "Aurora2CBAB212",
            "Endpoint.Port",
          ],
        },
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "DBSG9487F3EF",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "DBSG9487F3EF",
            "GroupId",
          ],
        },
        "ToPort": Object {
          "Fn::GetAtt": Array [
            "Aurora2CBAB212",
            "Endpoint.Port",
          ],
        },
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "DBSGtoTargetDBSG2D7B6556IndirectPortCDBCE918": Object {
      "Properties": Object {
        "Description": "Allow connections to the database Cluster from the Proxy",
        "DestinationSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "DBSG9487F3EF",
            "GroupId",
          ],
        },
        "FromPort": Object {
          "Fn::GetAtt": Array [
            "Aurora2CBAB212",
            "Endpoint.Port",
          ],
        },
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "DBSG9487F3EF",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "ToPort": Object {
          "Fn::GetAtt": Array [
            "Aurora2CBAB212",
            "Endpoint.Port",
          ],
        },
      },
      "Type": "AWS::EC2::SecurityGroupEgress",
    },
    "TargetAuroraSecret852055C73fdaad7efa858a3daf9490cf0a702aeb": Object {
      "Properties": Object {
        "Description": Object {
          "Fn::Join": Array [
            "",
            Array [
              "Generated by the CDK for stack: ",
              Object {
                "Ref": "AWS::StackName",
              },
            ],
          ],
        },
        "GenerateSecretString": Object {
          "ExcludeCharacters": " %+~\`#$&*()|[]{}:;<>?!'/@\\"\\\\",
          "GenerateStringKey": "password",
          "PasswordLength": 30,
          "SecretStringTemplate": "{\\"username\\":\\"dbAdmin\\"}",
        },
      },
      "Type": "AWS::SecretsManager::Secret",
    },
  },
}
`;
